### PR TITLE
Merge queue status page

### PR DESCRIPTION
**Pull request by dobbymoodge**

### CODE REVIEW

**Comment by dobbymoodge** [comment]
@danmcp @stevekuznetsov

File: merge_queue_overview [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+queue_name = PROPERTIES_LOCATION[/[^\w]+\.(.+)\.json/, 1]
+
+queue_records = if File.exists? MERGE_QUEUE_RECORD
+  JSON.parse(IO.read(MERGE_QUEUE_RECORD))
+else
+  $stderr.puts "Couldn't load merge queue records file #{MERGE_QUEUE_RECORD}"
+  exit 255
+end
+
+LOGO_URL = 'https://openshift.redhat.com/app/assets/openshift_logo_wide_blk250.png'
```
Comment (danmcp) [inline merge_queue_overview]: Needs to be in config

File: templates/merge_queue_overview.erb [hunk 2]
```diff
@@ -0,0 +1,80 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+<title><%= queue_name %> Merge Queue Overview</title>
+<script type='text/javascript' src='https://ajax.googleapis.com/ajax/libs/jquery/1.8.1/jquery.min.js?ver=3.8.1'></script>
+<link href="https://netdna.bootstrapcdn.com/bootstrap/3.1.1/css/bootstrap.min.css" type="text/css" rel="stylesheet" media="screen">
+<script src="https://netdna.bootstrapcdn.com/bootstrap/3.1.1/js/bootstrap.min.js"></script>
+<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,300,600,800,700' rel='stylesheet' type='text/css'>
+<link href="stylesheets/trello.css" rel="stylesheet" type="text/css">
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: Don't like the dep on trello.css.

File: test_pull_requests [hunk 3]
```diff
@@ -1928,6 +1933,11 @@ popd
               # Skip if it's not mergeable
               mergeable = is_mergeable?(id, repo)
               $stderr.puts "  Mergeable: #{mergeable}"
+              if req['merged']
+                pull_request_statuses[req['url']][:status] = :merged
```
Comment (danmcp) [inline test_pull_requests]: Shouldn't need all the different statuses anymore?

File: templates/merge_queue_overview.erb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    </div>
+
+  <div class="container">
+
+    <h2 class="visible-xs"><%= queue_name %> Merge Queue Overview</h2>
+
+<div class="container">
+<div class="table-responsive">
+<table class="table table-normal table-condensed">
+  <tr><th>Pull Request</th><th>Queue Position</th></tr>
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: Put queue position first and just call it position.

File: templates/merge_queue_overview.erb [hunk 5]
```diff
@@ -0,0 +1,82 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+<title><%= queue_name %> Merge Queue Overview</title>
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: upcase queue_name

File: templates/merge_queue_overview.erb (hunk 4 above)
Comment (danmcp) [inline templates/merge_queue_overview.erb]: Maybe even remove the column and just use 1: or 1 -

File: templates/merge_queue_overview.erb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  <div class="container">
+
+    <h2 class="visible-xs"><%= queue_name %> Merge Queue Overview</h2>
+
+<div class="container">
+<div class="table-responsive">
+<table class="table table-normal table-condensed">
+  <tr><th>Pull Request</th><th>Queue Position</th></tr>
+<%
+        queue_records.sort_by{ |pr,pd| pd['queue_pos'] || 9999 }.each do |pr_link, pr_data|
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: Why doesn't everything have a position or be in the process of merging.

File: .merge_queue_for_test_pull_requests.json.sample [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
+    "merge_queue_properties_location": "~/.test_pull_requests.json",
+    "logo_url": "https://openshift.redhat.com/app/assets/openshift_logo_wide_blk250.png"
```
Comment (danmcp) [inline .merge_queue_for_test_pull_requests.json.sample]: Change the image name for the sample.

File: templates/merge_queue_overview.erb [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+<table class="table table-normal table-condensed">
+  <tr><th>Running</th></tr>
+  <tr>
+  <td><a href='<%=pr_link%>'><%=pr_data['title']%></a></td>
+  </tr>
+</table>
+<% end %>
+
+<table class="table table-normal table-condensed">
+  <tr><th colspan="2">Merge Queue</th></tr>
```
Comment (dobbymoodge) [inline templates/merge_queue_overview.erb]: What about "Enqueued"?

Comment (danmcp) [inline templates/merge_queue_overview.erb]: +1

Comment (dobbymoodge) [inline templates/merge_queue_overview.erb]: nevermind, Waiting is better

File: .merge_queue_for_test_pull_requests.json.sample (hunk 7 above)
Comment (dobbymoodge) [inline .merge_queue_for_test_pull_requests.json.sample]: fixed

File: README.md [hunk 9]
```diff
@@ -53,6 +53,21 @@ Utility for serially testing and merging pull requests in conjunction with Jenki
 
  * Run `test_pull_requests` as a Jenkins or `cron` job.  Note that GitHub is rate limited but typical projects can still run this script every few mins without running out of requests.
 
+### Setup for merge_queue_overview
+ * Add `merge_queue_overview` to your Jenkins system and make sure it's executable
+ * Add `.merge_queue_for_test_pull_requests.json` to `$JENKINS_HOME`
```
Comment (danmcp) [inline README.md]: Let's rename this.  Maybe .merge_queue_overview.json

File: README.md [hunk 10]
```diff
@@ -53,6 +53,21 @@ Utility for serially testing and merging pull requests in conjunction with Jenki
...
+ * Create the `merge_queue_records` directory under `$JENKINS_HOME` and make sure its writable by the `jenkins` user.
```
Comment (danmcp) [inline README.md]: it's

File: README.md (hunk 9 above)
Comment (dobbymoodge) [inline README.md]: It needs to be distinct per merge queue

File: README.md (hunk 10 above)
Comment (dobbymoodge) [inline README.md]: fixed

File: README.md (hunk 9 above)
Comment (dobbymoodge) [inline README.md]: But for the example config, that's fine

File: README.md [hunk 11]
```diff
@@ -53,6 +53,21 @@ Utility for serially testing and merging pull requests in conjunction with Jenki
 
  * Run `test_pull_requests` as a Jenkins or `cron` job.  Note that GitHub is rate limited but typical projects can still run this script every few mins without running out of requests.
 
+### Setup for merge_queue_overview
+ * Add `merge_queue_overview` to your Jenkins system and make sure it's executable
+ * Add `.merge_queue_overview.json` to `$JENKINS_HOME`
```
Comment (danmcp) [inline README.md]: I don't love the name with _for for the one we are using either.  Also, if we have a file per set of config, let's add a name to it so you don't see test_pull_requests in the html name.

File: test_pull_requests [hunk 12]
```diff
@@ -2043,27 +2048,34 @@ popd
...
           if bot_comment
             fields = extract_bot_comment_fields(bot_comment['body'], settings)
             if (waiting_in_queue_state?(fields[:state]))
               skipped_count_branch = skipped_count[branch] ? skipped_count[branch] : skipped_count['*']
               skipped_count_branch[settings['name']] = 0 if skipped_count_branch[settings['name']].nil?
               skipped_count_branch[settings['name']] += 1
               queued_comment = compose_bot_comment(settings['test_prefix'], :state => :wait_in_queue, :content => waiting_in_queue_comment_segment(skipped_count_branch[settings['name']].to_s))
-              create_or_update_comment(req['number'], req['base']['repo']['name'], settings['test_prefix'], queued_comment , comments)
-              $stderr.puts "  Pull ##{req['number']} in repo '#{req['base']['repo']['name']}' is at build position ##{skipped_count_branch[settings['name']]}"
+              pull_request_statuses[:enqueued][req['url']][:title] = req['title']
```
Comment (stevekuznetsov) [inline test_pull_requests]: `req['url']` here is the API endpoint

Comment (stevekuznetsov) [inline test_pull_requests]: you want `html_url`

Comment (dobbymoodge) [inline test_pull_requests]: Thanks, fixed in latest push

File: templates/merge_queue_overview.erb [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    </div>
+
+  <div class="container">
+
+    <h2 class="visible-xs"><%= queue_name %> Merge Queue Overview</h2>
+
+<div class="container">
+<div class="table-responsive">
+<%
+  queue_records['running'].each do |pr_link, pr_data|
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: There is another case here where you don't have something running but you do have PRs waiting.  Namely if the previous merge job is running but the pr it's testing is already merged.  Should put some sort of message here for that case.  Ex: Waiting for previous merge job to finish...

File: templates/merge_queue_overview.erb [hunk 14]
```diff
@@ -0,0 +1,93 @@
+<!DOCTYPE html>
```
Comment (danmcp) [inline templates/merge_queue_overview.erb]: Needs a page icon

File: README.md (hunk 11 above)
Comment (dobbymoodge) [inline README.md]: fixed

**Comment by dobbymoodge** [comment]
@danmcp ready for merge, I do believe

**Pull request merged**